### ISSUE TITLE
`msgpack` failure

### ISSUE THREAD

**Issue by bennn** [racket/typed-racket#625]
cc @samth

### Description

The `msgpack` package fails to build with the current TR release candidate. The source compiles, but test programs written in `#lang racket` do not. For example:

```
#lang racket/base
(require msgpack)
unpack
```

Error message:

```
foo.rkt:3:0: Type Checker: could not convert type to a contract;
 required a flat contract but generated a chaperone contract
  identifier: unpack
  type: (-> Input-Port Packable)
  in: unpack
```

The type for unpack is:

```
(: unpack (-> Input-Port
              (U Void Boolean Integer Real String Bytes (Vectorof Packable) (Listof Packable) (HashTable Packable Packable) Ext)))
```

where `Packable`  is:

```
(define-type Packable
  (U Void
     Boolean
     Integer
     Real
     String
     Bytes
     (Vectorof Packable)
     (Listof Packable)
     (HashTable Packable Packable)
     Ext))
```

### Minimal Example

The TR error message seems to be due to the `HashTable` type in `Packable`; here's a tiny program the raises a similar exception:

```
#lang racket/base

(module t typed/racket/base
  (provide f)

  (define-type T
    (HashTable (Vectorof Integer) Symbol))

  (: f (-> Symbol T))
  (define (f x)
    (hash '#(1) x)))

(require 't)

f
```

Error message (on Racket v6.10.1 or earlier):

```
bar.rkt:15:0: Type Checker: could not convert type to a contract;
 required a flat contract but generated a chaperone contract
  identifier: f
  type: (-> Symbol T)
  in: f
```

The program runs if the type for `f` is `(-> Symbol (U Integer T))` because that type gets converted to a contract like `(-> Symbol any/c)`

### Diagnosis

Since `(hash (vector 1) 'a)` is valid, the problem is likely the constraint in [`static-contracts/combinators/structural.rkt`](https://github.com/racket/typed-racket/blob/master/typed-racket-lib/typed-racket/static-contracts/combinators/structural.rkt) that hashtable keys must be flat:

```
  ((hash/sc (#:invariant #:flat) (#:invariant)) hash/c #:chaperone)
```

**Comment by pnwamk** [comment]
Was the motivation for requiring flat key contracts perhaps to not change the behavior of `eq?` based hash tables?

**Comment by samth** [comment]
I think we can fix that by distinguishing `equal?` based tables. Will that fix `msgpack`?

**Comment by bennn** [comment]
After removing `#:flat`, `msgpack` builds and all `raco test -c msgpack` tests pass

**Comment by pnwamk** [comment]
@samth what do you mean distinguish in this case? As in at the type level? (I assume not since that would be non-trivial and your talking about fixing for release) Or something else?

**Comment by samth** [comment]
Ok great. We need to improve the contract generated for `hash/sc` if the keys aren't flat so that it doesn't have a weird contract-system error when given an `eq?` or `eqv?` based hash, and then we should be good to go.

**Comment by bennn** [comment]
@samth I think I understand that, will send a PR soon

**Comment by pnwamk** [comment]
Maybe I'll understand once the PR is ready XD

**Comment by samth** [comment]
See the docs for `hash/c`

**Comment by pnwamk** [comment]
I've read them -- I'll just wait for the PR and ask if needed their 👍

**Comment by bennn** [comment]
closed by #626
